semantic-convention-utils/src/main/java/org/hypertrace/semantic/convention/utils/http/HttpSemanticConventionUtils.java

```diff
@@ -196,4 +257,257 @@ public static boolean isValidUrl(String url) {
+    if (event.getAttributes() == null || event.getAttributes().getAttributeMap() == null) {
+      return Optional.empty();
+    }
+
+    Map<String, AttributeValue> attributeValueMap = event.getAttributes().getAttributeMap();
+    for (String url : FULL_URL_ATTRIBUTES) {
+      if (attributeValueMap.get(url) != null) {
+        String urlVal = attributeValueMap.get(url).getValue();
+        if (!StringUtils.isBlank(urlVal) && isValidUrl(urlVal)) {
+          if (isAbsoluteUrl(urlVal)) {
+            return Optional.of(urlVal);
+          } else {
+            Optional<String> url2 = getHttpUrlForOTelFormat(attributeValueMap);
+            return url2.isPresent() ? url2 : Optional.of(urlVal);
+          }
+        }
+      }
+    }
+
+    return getHttpUrlForOTelFormat(attributeValueMap);
+  }
+
+  public static Optional<String> getHttpQueryString(Event event) {
+
+    Optional<String> queryString = getHttpQueryStringUtil(event);
+    Optional<String> url = getHttpUrlUtil(event);
+    if (url.isPresent() && queryString.isEmpty()) {
+      try {
+        return Optional.ofNullable(getNormalizedUrl(url.get()).getQuery());
+      } catch (MalformedURLException e) {
+        e.printStackTrace();
+      }
+    }
+    return queryString;
+  }
+
+  private static Optional<String> getHttpQueryStringUtil(Event event) {
+    if (event.getAttributes() == null || event.getAttributes().getAttributeMap() == null) {
+      return Optional.empty();
+    }
+
+    Map<String, AttributeValue> attributeValueMap = event.getAttributes().getAttributeMap();
+    for (String query_string : QUERY_STRING_ATTRIBUTES) {
+      if (attributeValueMap.get(query_string) != null) {
+        return Optional.of(attributeValueMap.get(query_string).getValue());
+      }
+    }
+    return Optional.empty();
+  }
+
+  public static Optional<Integer> getHttpRequestSize(Event event) {
+    if (event.getAttributes() == null || event.getAttributes().getAttributeMap() == null) {
+      return Optional.empty();
+    }
+
+    Map<String, AttributeValue> attributeValueMap = event.getAttributes().getAttributeMap();
+    for (String reqsa : REQUEST_SIZE_ATTRIBUTES) {
+      if (attributeValueMap.get(reqsa) != null) {
+        return Optional.of(Integer.parseInt(attributeValueMap.get(reqsa).getValue()));
+      }
+    }
+    return Optional.empty();
+  }
+
+  public static Optional<Integer> getHttpResponseSize(Event event) {
+    if (event.getAttributes() == null || event.getAttributes().getAttributeMap() == null) {
+      return Optional.empty();
+    }
+
+    Map<String, AttributeValue> attributeValueMap = event.getAttributes().getAttributeMap();
+    for (String rsa : RESPONSE_SIZE_ATTRIBUTES) {
+      if (attributeValueMap.get(rsa) != null) {
+        return Optional.of(Integer.parseInt(attributeValueMap.get(rsa).getValue()));
+      }
+    }
+    return Optional.empty();
+  }
+
+  public static Optional<String> getPathFromUrlObject(String urlPath) {
```

Comment:
why public?

Reply:
it is used in a test

Reply:
Making functions public just to be used in unit test is a bad design. Basically, we are making a function public, which would can now be accessible by anyone using the utils, though it wasn't supposed to be public actually

Reply:
I have 2 utility functions in this file. Then should I move them to a different file and then import and use?

Reply:
I don't think so. Basically, the question being

Do we want these functions to be public? 

- If yes, then there is no issue in unit testing them directly, since we should always unit test the public functions
- If no, since this is a private function, you should not unit test it **directly**. The unit test for the caller function  should cover the unit tests for the callee function. 

In your case, it would be the unit tests of the function `getHttpPath` (public) should cover test cases for `getHttpPathUtil` (private) and `getPathFromUrlObject` (private)

Reply:
Made them package private

Reply:
Making them package private (`protected`) definitely reduces the issue, but it's still the same issue. 

Any other classes in the same package would still be able to access the `protected` method, though the intention was not to expose them outside from the current scope of the class